File: Sum_Root_to_Leaf_Numbers.py
'''
Time Complexity - O(n)
Space Complexity - O(h)
'''

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        self.sum = 0
        return self.helper(root, 0)

    def helper(self, root, currSum):
        if root == None:
            return 0
        currSum = currSum*10 + root.val
        if root.left == None and root.right == None:
            self.sum += currSum
        self.helper(root.left, currSum)
        self.helper(root.right, currSum)
        return self.sum

File: SumRootToLeaf/SumRootToLeafNumbers.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

 /* Time Complexity : O(n)
 * 	n - size of the tree */
/* Space Complexity : O(h) 
 * 	h - height of the tree */

 // retun type (int) based recursion

class Solution {
    public int sumNumbers(TreeNode root) {
        return helper(root, 0);
    }

    private int helper(TreeNode root, int currNum){
        if(root == null) return 0;
        //Number at node
        currNum = currNum * 10 + root.val;
        //check for leaf
        if(root.left == null && root.right == null){
            return currNum;
        }
        int left = helper(root.left, currNum);
        int right = helper(root.right, currNum);

        return left + right;
    }
}

// void based recursion

/*
class Solution {
    int result;
    public int sumNumbers(TreeNode root) {
        helper(root, 0);
        return result;
    }

    private void helper(TreeNode root, int currNum){
        if(root == null) return;
        //Number at node
        currNum = currNum * 10 + root.val;
        //check for leaf
        if(root.left == null && root.right == null){
            result += currNum;
        }
        helper(root.left, currNum);
        helper(root.right, currNum);
    }
}
*/